src/compute-client/src/controller.rs

```diff
@@ -889,19 +898,219 @@ impl<T: ComputeControllerTimestamp> CollectionState<T> {
+                    IntrospectionType::ComputeMaterializedViewRefreshes,
+                    vec![(retraction, -1)],
+                );
+            });
+    }
+}
+
+/// Information needed to compute introspection updates for a REFRESH materialized view when the
+/// write frontier advances.
+#[derive(Debug)]
+struct RefreshIntrospectionState<T> {
+    // Immutable properties of the MV
+    refresh_schedule: RefreshSchedule,
+    initial_as_of: Antichain<T>,
+    // Refresh state
+    next_refresh: Datum<'static>,
+    last_completed_refresh: Datum<'static>,
+}
+
+impl<T> RefreshIntrospectionState<T> {
+    /// Return a `Row` reflecting the current refresh introspection state.
+    pub fn row_for_collection(&self, collection_id: GlobalId) -> Row {
+        Row::pack_slice(&[
+            Datum::String(&collection_id.to_string()),
+            self.last_completed_refresh,
+            self.next_refresh,
+        ])
+    }
+}
+
+impl<T: ComputeControllerTimestamp> RefreshIntrospectionState<T> {
+    pub fn new(
+        refresh_schedule: RefreshSchedule,
+        initial_as_of: Antichain<T>,
+        upper: Antichain<T>,
+    ) -> RefreshIntrospectionState<T> {
+        let mut self_ = RefreshIntrospectionState {
+            refresh_schedule: refresh_schedule.clone(),
+            initial_as_of: initial_as_of.clone(),
+            next_refresh: Datum::Null,
+            last_completed_refresh: Datum::Null,
+        };
+        self_.frontier_update(&upper);
+        self_
+    }
+
+    pub fn frontier_update(&mut self, write_frontier: &Antichain<T>) {
+        if write_frontier.is_empty() {
+            self.last_completed_refresh =
+                if let Some(last_refresh) = self.refresh_schedule.last_refresh() {
+                    last_refresh.into()
+                } else {
+                    // If there is no last refresh, then we have a `REFRESH EVERY`, in which case
+                    // the saturating roundup puts a refresh at the maximum possible timestamp.
+                    T::maximum().into()
+                };
+            self.next_refresh = Datum::Null;
+        } else {
+            if timely::PartialOrder::less_equal(write_frontier, &self.initial_as_of) {
+                // We are before the first refresh.
+                self.last_completed_refresh = Datum::Null;
+                let initial_as_of = self.initial_as_of.as_option().expect(
+                    "initial_as_of can't be [], because then there would be no refreshes at all",
+                );
+                let first_refresh = initial_as_of
+                    .round_up(&self.refresh_schedule)
+                    .expect("sequencing makes sure that REFRESH MVs always have a first refresh");
+                soft_assert_or_log!(
+                    first_refresh == *initial_as_of,
+                    "initial_as_of should be set to the first refresh"
+                );
+                self.next_refresh = first_refresh.into();
```

Comment:
We do expect frontier advancements before the initial as_of, because the initial as_of belongs to the storage collection, but the dataflow as_of is usually earlier than that in case of REFRESH MVs, to allow for warmup before refreshes, see [here](https://github.com/MaterializeInc/materialize/blob/4f91d6612f7f4843d5fbc572dfbf91c477de8a9d/src/adapter/src/coord/sequencer/inner/create_materialized_view.rs#L759-L763).

(Btw. we couldn't rely on not seeing frontier advances before even the dataflow as_of, as [the contract for the dataflow as_of is that data won't come before it, but frontier advancements can happen before it](https://github.com/MaterializeInc/materialize/blob/4f91d6612f7f4843d5fbc572dfbf91c477de8a9d/src/compute-types/src/dataflows.rs#L63-L65).)